src/lib/util/compat.ml

```diff
@@ -0,0 +1,131 @@
+(*                                                                        *)
+(*     This file is distributed under the terms of OCamlPro               *)
+(*     Non-Commercial Purpose License, version 1.                         *)
+(*                                                                        *)
+(*     As an exception, Alt-Ergo Club members at the Gold level can       *)
+(*     use this file under the terms of the Apache Software License       *)
+(*     version 2.0.                                                       *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Xavier Leroy, projet Cristal, INRIA Rocquencourt           *)
+(*                                                                        *)
+(*   Copyright 1996 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     More details can be found in the directory licenses/               *)
+(*                                                                        *)
+(**************************************************************************)
+
+module List = struct
+  include List
+
+  let is_empty = function
+    | [] -> true
+    | _ -> false
+
+  let rec equal eq l1 l2 =
+    match l1, l2 with
+    | [], [] -> true
+    | [], _::_ | _::_, [] -> false
+    | a1::l1, a2::l2 -> eq a1 a2 && equal eq l1 l2
+
+  let rec compare cmp l1 l2 =
+    match l1, l2 with
+    | [], [] -> 0
+    | [], _::_ -> -1
+    | _::_, [] -> 1
+    | a1::l1, a2::l2 ->
+      let c = cmp a1 a2 in
+      if c <> 0 then c
+      else compare cmp l1 l2
+
+  let rec find_map f = function
+    | [] -> None
+    | x :: l ->
+      begin match f x with
+        | Some _ as result -> result
+        | None -> find_map f l
+      end
+
+  let fold_left_map f accu l =
+    let rec aux accu l_accu = function
+      | [] -> accu, rev l_accu
+      | x :: l ->
+        let accu, x = f accu x in
+        aux accu (x :: l_accu) l in
+    aux accu [] l
+end
+
+module Bytes = struct
+  include Bytes
```

Comment:
I would argue it would be better to do `open Bytes` at the beginning of the module, and `include Bytes` at the end (with `[@@@ocaml.warning "-32-33"]` at the top of the file). This way, we would use the stdlib version when it is available and only fall back to our version if its not.

(Same for the `List`, `Array`, etc. in this module)

Reply:
I was looking for such trick in fact, thanks :)